### PR TITLE
Feature/sc 121837/support local flash without internet access

### PR DESCRIPTION
**Pull request by hugomontero**
## Description
This PR will allow the use of flashing device os offline.
When the user downloads an specific device os version it will be cached in a file, then if the user tries to download it again without internet access, then the `cli` will looks into the cached file to try to flash the device.

## How to Test

1. Pull down the branch: `git pull && git checkout feature/sc-121837/support-local-flash-without-internet-access`
2. Install dependencies: `npm i`
3. Run tests: `npm test`
4. Attempt to run `steps with internet access` 
5. Attempt to run `steps without internet access`
6. Remove the `device-os-version-cached.json` from `~/.particle/device-os-flash/binaries`
7. Attempt to run `steps without internet access`
8. Re-run step 4
9. Delete the binaries from `~/.particle/device-os-flash/binaries/${version}/${platform}`
10. Attempt to run `steps without internet access`

**Steps with internet access:**
* Update your device to:
    * latest: `npm start -- update`
    * specific target: `npm start -- update --target 5.0.0`

**Steps Without internet access**
* Disconnect the client from internet
* Update your device to:
    * latest: `npm start -- update`
    * specific target: `npm start -- update --target 5.0.0`

**outcome** 
* Tests should pass
* Steps 4 and 5 should flash the device without issues and should create a new file in `~/.particle/device-os-flash/binaries` named `device-os-version-cached.json`
* Step 7 and 10 should show a message indicating: `Device OS version not found in cache for platform: ${platformId} version: ${version} and there was an internet connection error`

## Related Issues / Discussions

Story details 👉  https://app.shortcut.com/particle/story/121837/support-local-flash-without-internet-access

## Completeness

- [x] User is totes amazing for contributing!
- [x] Contributor has signed [CLA](https://docs.google.com/a/particle.io/forms/d/1_2P-vRKGUFg5bmpcKLHO_qNZWGi5HKYnfrrkd-sbZoA/viewform)
- [x] Problem and solution clearly stated
- [x] Tests have been provided
- [x] Docs have been updated
- [x] CI is passing

### CODE REVIEW

File: src/lib/device-os-version-cache.js [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+		} catch (error) {
+			throw new Error(`Error opening or creating the cache file: ${error.message}`);
+		}
+	}
+
+	async _saveCache() {
+		const particleDir = ensureFolder();
+		const filePath =  path.join(particleDir, CACHE_DIR_NAME, CACHE_FILE_NAME);
+		try {
+			await fs.writeJson(filePath, this.cache);
```
Comment (monkbroc) [inline src/lib/device-os-version-cache.js:61]: Did you test when the directory doesn't exist? You could use [outputJson](https://github.com/jprichardson/node-fs-extra/blob/47bc43d036a0aa0b0429a478da78122793b42bdb/docs/outputJson.md#outputjsonfile-object-options-callback) to create the directory if it doesn't exist

File: src/lib/device-os-version-cache.js [hunk 2]
```diff
@@ -0,0 +1,72 @@
+const path = require('path');
+const { ensureFolder } = require('../../settings');
+const fs = require('fs-extra');
+const CACHE_FILE_NAME = '.device-os-version.cache';
```
Comment (monkbroc) [inline src/lib/device-os-version-cache.js:4]: Since the cache file is JSON, make the extension `.json`. No need for a dot at the start to hide the file.

**Review by monkbroc — APPROVED** [review]
I tested it. It works great now and we have a repeatable pattern for offline usage of the API!

File: src/lib/particle-cache.js [hunk 3]
```diff
@@ -0,0 +1,40 @@
+const path = require('path');
+const settings = require('../../settings');
+const fs = require('fs-extra');
+const crypto = require('crypto');
+
+
+class ParticleCache {
+	constructor() {
+		this.path = path.join(settings.findHomePath(), '.particle', '.cache');
```
[the remark this answers is not in the document]
Comment (hugomontero) [reply src/lib/particle-cache.js]: Fixed here! https://github.com/particle-iot/particle-cli/pull/678/commits/30ad319ca7c07cbff1d95f5b51c6adc38a70fbe6

**Pull request merged**